packages/core/src/electron-main/electron-main-application.ts

```diff
@@ -0,0 +1,357 @@
+
+    onQuit?(event: electron.Event, exitCode: number): void;
+}
+
+export interface ElectronBrowserWindowOptions extends electron.BrowserWindowConstructorOptions {
+    isMaximized?: boolean,
+}
+
+const WindowState = 'windowstate';
+interface WindowState {
+    isMaximized?: boolean
+    height: number
+    width: number
+    x: number
+    y: number
+}
+export interface ElectronMainApplicationStorage {
+    [WindowState]: WindowState;
+}
+
+export const TheiaApplicationName = Symbol('TheiaApplicationName');
+export type TheiaApplicationName = string;
+
+export const TheiaBackendMainPath = Symbol('TheiaBackendMainPath');
+export type TheiaBackendMainPath = string;
+
+export const TheiaIndexHtmlPath = Symbol('TheiaIndexHtmlPath');
+export type TheiaIndexHtmlPath = string;
+
+@injectable()
+export class ElectronMainApplication {
+
+    @inject(TheiaApplicationName)
+    protected readonly applicationName: TheiaApplicationName;
+
+    @inject(TheiaBackendMainPath)
+    protected readonly mainPath: TheiaBackendMainPath;
+
+    @inject(TheiaIndexHtmlPath)
+    protected readonly indexHtml: TheiaIndexHtmlPath;
+
+    @inject(ContributionProvider) @named(ElectronMainApplicationContribution)
+    protected readonly contributions: ContributionProvider<ElectronMainApplicationContribution>;
+
+    // tslint:disable-next-line:no-any
+    protected readonly storage = new ElectronStorage<ElectronMainApplicationStorage>();
+
+    protected readonly backendPortDeferred = new Deferred<number>();
+    /**
+     * Set when the NodeJS backend has started and dispatched the port it is running on.
+     */
+    protected readonly backendPort = this.backendPortDeferred.promise;
+
+    async start(app: electron.App): Promise<void> {
+        await this.bindApplicationEvents(app);
+        this.startBackend(app); // no await here, let it be concurrent.
+        try {
+            await Promise.all(this.contributions.getContributions()
+                .map(contribution => contribution.onStart && contribution.onStart(app)));
+        } catch (error) {
+            console.error(error);
+            app.exit(1);
+            return;
+        }
+        this.ready(app);
+    }
+
+    protected async ready(app: electron.App): Promise<void> {
+        const infos = await app.whenReady();
+        this.bindIpcEvents();
+        await this.setTempMenu();
+        await Promise.all(this.contributions.getContributions()
+            .map(contribution => contribution.onReady && contribution.onReady(infos)));
+        await this.openWindow();
+    }
+
+    protected canPreventStop = true;
+    protected async beforeQuit(event: electron.Event): Promise<void> {
+        if (this.canPreventStop) {
+            // Pause the stop.
+            event.preventDefault();
+            let preventStop = false;
+            // Ask all opened windows whether they want to prevent the \`close\` event or not.
+            for (const window of electron.BrowserWindow.getAllWindows()) {
```

Comment:
I rebased using gitpod, but I think I should test that locally, really don't know where to move this code right now.